subsys/net/lib/sockets/sockets_service.c

```diff
@@ -0,0 +1,245 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(net_sock_svc, CONFIG_NET_SOCKETS_LOG_LEVEL);
+
+#include <zephyr/kernel.h>
+#include <zephyr/net/socket_service.h>
+#include <zephyr/posix/sys/eventfd.h>
+
+static int init_socket_service(void);
+static bool init_done;
+
+static K_MUTEX_DEFINE(lock);
+static K_CONDVAR_DEFINE(wait_start);
+
+static struct service {
+	struct zsock_pollfd *events;
+	int count;
+} ctx;
+
+void net_socket_service_foreach(net_socket_service_cb_t cb, void *user_data)
+{
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		cb(svc, user_data);
+	}
+}
+
+static void cleanup_svc_events(struct net_socket_service_desc *svc)
+{
+	for (int i = svc->idx; i < svc->idx + svc->pev_len; i++) {
+		for (int j = 0; j < svc->pev_len; j++) {
+			if (ctx.events[i].fd >= 0 &&
+			    ctx.events[i].fd == svc->pev[j].event.fd) {
+				ctx.events[i].fd = -1;
+				svc->pev[j].event.fd = -1;
+				svc->pev[j].event.events = 0;
+			}
+		}
+	}
+}
+
+int z_impl_net_socket_service_register(struct net_socket_service_desc *service,
+				       struct zsock_pollfd *fds, int len,
+				       void *user_data)
+{
+	int i, ret = -ENOENT;
+
+	k_mutex_lock(&lock, K_FOREVER);
+
+	if (!init_done) {
+		(void)k_condvar_wait(&wait_start, &lock, K_FOREVER);
+	}
+
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		if (svc == service) {
+			if (fds == NULL) {
+				cleanup_svc_events(svc);
+			} else {
+				if (len > svc->pev_len) {
+					NET_DBG("Too many file descriptors, max is %d for service %p",
+						svc->pev_len, svc);
+					ret = -ENOMEM;
+					break;
+				}
+
+				for (i = 0; i < len; i++) {
+					svc->pev[i].event = fds[i];
+					svc->pev[i].user_data = user_data;
+				}
+
+				for (i = svc->idx; i < svc->idx + svc->pev_len; i++) {
+					for (int j = 0; j < svc->pev_len; j++) {
+						ctx.events[i] = svc->pev[i].event;
+					}
+				}
+			}
+
+			/* Tell the thread to re-read the variables */
+			eventfd_write(ctx.events[0].fd, 1);
+			ret = 0;
+
+			break;
+		}
+	}
+
+	k_mutex_unlock(&lock);
+
+	return ret;
+}
+
+static struct net_socket_service_desc *find_svc_and_event(
+	struct zsock_pollfd *pev,
+	struct net_socket_service_event **event)
+{
+	STRUCT_SECTION_FOREACH(net_socket_service_desc, svc) {
+		for (int i = 0; i < svc->pev_len; i++) {
+			if (svc->pev[i].event.fd == pev->fd) {
+				*event = &svc->pev[i];
+				return svc;
+			}
+		}
+	}
+
+	return NULL;
+}
+
+static int trigger_work(struct zsock_pollfd *pev)
+{
+	struct net_socket_service_event *event;
+	struct net_socket_service_desc *svc;
+	int ret;
+
+	svc = find_svc_and_event(pev, &event);
+	if (svc == NULL) {
+		return -ENOENT;
+	}
+
+	if (svc->work_q != NULL) {
```

Comment:
Shouldn't we call the registred callback directly from the services thread? I see a potential race here - we do submit the work to execute, but don't read the data (in case POLLIN was monitored) from the socket until the work executes. If the workqueue thread has lower prio than the services thread, we may end up busy looping  (poll() will keep reporting POLLIN as the work item didn't have a chance to execute).

Reply:
> Shouldn't we call the registred callback directly from the services thread?

I originally planned to do that, but it felt more natural to use the existing APIs and resources in the system. So using a work queue does the same thing. But we can certainly change this, it is just to change one function call.

When using workqueue(s), we can serve (at least theoretically) multiple socket services at the "same" time and the priority of the thread using the services api determines which one gets called first. If we have a callback directly from services thread, then there is no such thing available.

Reply:
> I see a potential race here - we do submit the work to execute, but don't read the data (in case POLLIN was monitored) from the socket until the work executes.

Note that `k_work_submit()` does nothing if the work is already pending. It is indeed possible that we might busyloop in the service thread if the service user does not "run" fast enough. I tried to mitigate this problem by setting the service thread priority to lowest application thread priority.

Reply:
There seems to an issue when calling the "callback" via the workqueue as it might be possible (depending on different thread priorities) that we start to loop through poll() if the callback does not get run properly. So it might be better if we call the callback directly (synchronously) without using the workqueue. I will investigate this more.

Reply:
> There seems to an issue when calling the "callback" via the workqueue as it might be possible (depending on different thread priorities) that we start to loop through poll() if the callback does not get run properly. So it might be better if we call the callback directly (synchronously) without using the workqueue. I will investigate this more.

Left the async call but fixed the code so that the callback work is only called once. Fixed also the tests that were not working properly with native_sim.

Reply:
To be honest I'm not personally convinced about this async approach. Having multiple work queues kind of busts the idea the socket services were created for (saving resources), so most of the handlers will likely end up on a system workqueue anyway. Plus, having the handlers to be called directly from the services thread allows to reuse the services stack and kind of finetune its size for the worst case scenario (I imagine all of the registered services would need to allocate buffers for socket send/recv calls and a common stack area seemed like a good place for this instead of using static/heap/custom stack memory).

But it's just my personal opinion, no strong push from my side to change it, if that's the preferred way.

Reply:
Personally I prefer the async way of doing things although it certainly complicates things. But I can see that sync way is also possible here. I think we can support both methods quite easily and let the user select how the callback should be called. I will propose a new version that supports both methods.

Reply:
Yes - sync should be a relatively trivial wrapper around async ... or at least I thought it was... 🤔